Potential racing issue on agent/activity creation, vs multiple statements issues at once

We are getting into an issue where an agent or activity is attempted to be created more than once, by sending a bunch of statements asynchronously.

From our tests, we assume that a check is made to see if the agent/activity should be created, but by the time the agent/activity is actually created, a few requests to create it has gone through to a queue.

This causes statements to get lost as the system refuses to create a duplicate agent/activity:

```
Exception while processing: /xAPI/statements
2015-11-10 15:35:25,771 [ERROR] lrs.views: duplicate key value violates unique constraint "lrs_activity_activity_id_key"
DETAIL:  Key (activity_id)=(http://blahblah/someactivity) already exists.
Traceback (most recent call last):
  File "/home/adllrs/ADL_LRS/lrs/views.py", line 619, in handle_request
    return processors[path][req_dict['method']](req_dict)
  File "/home/adllrs/ADL_LRS/lrs/util/req_process.py", line 124, in statements_post
    stmt_responses = process_body(body, auth, req_dict['headers']['X-Experience-API-Version'], req_dict.get('payload_sha2s', None))
  File "/home/adllrs/ADL_LRS/lrs/util/req_process.py", line 57, in process_body
    return [process_statement(st, auth, version, payload_sha2s) for st in stmts]
  File "/home/adllrs/ADL_LRS/lrs/util/req_process.py", line 50, in process_statement
    st = StatementManager(stmt, auth, payload_sha2s).model_object
  File "/home/adllrs/ADL_LRS/lrs/managers/StatementManager.py", line 15, in __init__
    self.populate(auth_info, stmt_data, payload_sha2s)
  File "/home/adllrs/ADL_LRS/lrs/managers/StatementManager.py", line 166, in populate
    self.build_statement_object(auth_info, stmt_data)
  File "/home/adllrs/ADL_LRS/lrs/managers/StatementManager.py", line 152, in build_statement_object
    define=auth_info['define']).Activity
  File "/home/adllrs/ADL_LRS/lrs/managers/ActivityManager.py", line 7, in __init__
    self.populate(data)
  File "/home/adllrs/ADL_LRS/lrs/managers/ActivityManager.py", line 63, in populate
    self.Activity = Activity.objects.create(activity_id=activity_id, authority=self.auth)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 137, in create
    return self.get_query_set().create(**kwargs)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 377, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 463, in save
    self.save_base(using=using, force_insert=force_insert, force_update=force_update)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/base.py", line 551, in save_base
    result = manager._insert([self], fields=fields, return_id=update_pk, using=using, raw=raw)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/manager.py", line 203, in _insert
    return insert_query(self.model, objs, fields, **kwargs)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/query.py", line 1576, in insert_query
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/models/sql/compiler.py", line 910, in execute_sql
    cursor.execute(sql, params)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/backends/util.py", line 40, in execute
    return self.cursor.execute(sql, params)
  File "/home/adllrs/env/local/lib/python2.7/site-packages/django/db/backends/postgresql_psycopg2/base.py", line 52, in execute
    return self.cursor.execute(query, args)
IntegrityError: duplicate key value violates unique constraint "lrs_activity_activity_id_key"
DETAIL:  Key (activity_id)=(http://blahblah/someactivity) already exists.
```

On the other hand, if we make the same calls one after the other with a delay between them, we do not get these errors.

--- a/lrs/managers/ActivityManager.py
+++ b/lrs/managers/ActivityManager.py
@@ -1,3 +1,5 @@
+from django.db import IntegrityError
+
 from ..models import Activity
 
 class ActivityManager():
@@ -57,14 +59,21 @@ class ActivityManager():
         try:
             act = Activity.objects.get(activity_id=activity_id)
         except Activity.DoesNotExist:
-            act_created = True
             if self.define_permission:
-                # If activity DNE and can define - create activity with auth
-                self.Activity = Activity.objects.create(activity_id=activity_id, authority=self.auth)
                 can_define = True
+                # If activity DNE and can define - create activity with auth
+                try:
+                    self.Activity, act_created = Activity.objects.get_or_create(activity_id=activity_id, authority=self.auth)       
+                except IntegrityError:
+                    self.Activity = Activity.objects.get(activity_id=activity_id, authority=self.auth)
+                    act_created = False
             else:
                 # If activity DNE and cannot define - create activity without auth
-                self.Activity = Activity.objects.create(activity_id=activity_id)
+                try:
+                    self.Activity, act_created = Activity.objects.get_or_create(activity_id=activity_id)
+                except IntegrityError:
+                    self.Activity = Activity.objects.get(activity_id=activity_id)
+                    act_created = False
         # activity already exists
         else:
             self.Activity = act

--- a/lrs/models.py
+++ b/lrs/models.py
@@ -4,7 +4,7 @@ from datetime import datetime
 from jsonfield import JSONField
 
 from django_extensions.db.fields import UUIDField
-from django.db import models
+from django.db import models, IntegrityError
 from django.contrib.auth.models import User
 from django.core.files.storage import FileSystemStorage
 from django.utils.timezone import utc
@@ -104,8 +104,13 @@ class AgentManager(models.Manager):
                 created = False
             except IndexError:
                 # If DNE create the agent based off of kwargs (kwargs now includes account_homePage and account_name fields)
-                agent = Agent.objects.create(**kwargs)
-                created = True
+                try:
+                    agent = Agent.objects.create(**kwargs)
+                    created = True
+                except IntegrityError:
+                    # Try getting agent by IFP in ifp_dict
+                    agent = Agent.objects.filter(**ifp_dict)[0]
+                    created = False                    
 
             # For identified groups with members
             if is_group and has_member:
@@ -129,8 +134,12 @@ class AgentManager(models.Manager):
                     agent = Agent.objects.get(oauth_identifier=created_oauth_identifier)
                     created = False
                 except Agent.DoesNotExist:
-                    agent = Agent.objects.create(**kwargs)
-                    created = True
+                    try:
+                        agent = Agent.objects.create(**kwargs)
+                        created = True
+                    except IntegrityError:
+                        agent = Agent.objects.get(oauth_identifier=created_oauth_identifier)
+                        created = False
             # If oauth account is in second member
             elif 'account' in member[1] and 'OAuth' in member[1]['account']['homePage']:
                 created_oauth_identifier = "anongroup:%s-%s" % (member[1]['account']['name'], member[0]['mbox'])
@@ -138,8 +147,12 @@ class AgentManager(models.Manager):
                     agent = Agent.objects.get(oauth_identifier=created_oauth_identifier)
                     created = False
                 except Agent.DoesNotExist:
-                    agent = Agent.objects.create(**kwargs)
-                    created = True
+                    try:
+                        agent = Agent.objects.create(**kwargs)
+                        created = True
+                    except IntegrityError:
+                        agent = Agent.objects.get(oauth_identifier=created_oauth_identifier)
+                        created = False
             # Non-oauth anonymous group that has 2 members, one having an account
             else:
                 agent = Agent.objects.create(**kwargs)

--- a/lrs/utils/retrieve_statement.py
+++ b/lrs/utils/retrieve_statement.py
@@ -11,7 +11,7 @@ from django.db.models import Q
 
 from . import convert_to_utc
 from ..models import Statement, Agent
-from ..exceptions import NotFound, IDNotFoundError
+from ..exceptions import NotFound
 
 MORE_ENDPOINT = '/xapi/statements/more/'
 

